lib/matplotlib/figure.py

```diff
@@ -1113,9 +1113,29 @@ def legend(self, *args, **kwargs):
             # extra_args = extra_args[1:]
             pass
         transform = kwargs.pop('bbox_transform', self.transSubfigure)
+        outside = None
+        if 'loc' in kwargs:
+            # need to protect this in case loc was passed in
+            # positionally...
+            loc = kwargs.pop('loc', None)
+            # handle outside legends:
+            if isinstance(loc, str):
+                if loc.split()[0] == 'outside':
+                    loc = loc.replace('outside ', '')
+                    outside = loc.replace('center ', '')
+                    outside = outside.split()[0]
+                    locs = loc.split()
+                    if len(locs) > 1 and locs[0] in ('right', 'left'):
+                        # locs doesn't accept "left upper", etc, so swap
+                        if locs[0] != 'center':
+                            locs = locs[::-1]
+                        loc = locs[0] + ' ' + locs[1]
+
+            kwargs['loc'] = loc
         # explicitly set the bbox transform if the user hasn't.
         l = mlegend.Legend(self, handles, labels, *extra_args,
                            bbox_transform=transform, **kwargs)
+        l._outside = outside
```

Comment:
I'm not following the logic here. `l._outside` seems to dynamically add an attribute to the Legend. It's better to to pass this via `__init__` and document there what the type and expected values are.


I'm wondering whether the whole logic would be better moved in to `Legend.__init__`. On the downside, not every legend can cope with "outside", OTOH, if you don't hack this in via a dynamic attribute, you'd have a formal `outside` parameter in `Legend.__init__` and you'd have to cope with the outside case anyway. In that situation it seems simpler to make `Legend.__init__(... loc=...)` accept whatever is needed (i.e. also 'ouside upper left') and do all the logic in there.

Reply:
Except the loc keyword arg does error checking at the legend level, and legend doesn't know if it is being called as a figure legend or an axes legend.  I think it's better to keep figure.legend only code in the figure.legend method

Reply:
But the legend carries the information where it wants to be rendered. That's `loc` and now ammended with `_outside`.
By introducing "outside", positioning has become more complex. While we already have the case that the meaning of `loc` changes depending on `bbox_to_anchor`, we still got away with a cheap validation because the supported values are the same.

That's now getting a bit more complicated. I still think it's valuable to have all the validation logic in one place. We may need to consider making legend aware on whether it's positioning in a figure or Axes.

Anyway, the outside state should flow through a public API into Legend, otherwise downstream users/libraries will not be able to create their own outside `Legend`s.

Reply:
> Anyway, the outside state should flow through a public API into Legend, otherwise downstream users/libraries will not be able to create their own outside Legends.

Can you clarify what you have in mind here?  

This is a pretty over specified problem, where at your and @efiring request we are offering a different idiom for axes than figures for the *loc* keyword argument.  Now you are suggesting that somehow legend itself must parse that, and parse it differently whether it is a figure legend or an axes legend, information the legend does not currently have. 

That all seems very confusing to me, and very hard to document.  Either we have the grammar for _loc_ proposed here and parse it in the figure method and leave the general object mostly alone. Or the original proposal was to have an optional kwarg "outside" that was not made available to the axes method.  That seems cleaner to me, and less likely to lead to confusion, but...

This PR does the mechanical part of adjusting the constrained layout to accept the legend.  This is an oft requested feature, but it's not one I particularly plan to use.  So if there is going to be a bunch more API discussion I'd like to suggest that one of the two forms above be merged so the functionality is present and then a follow up PR can work out the API and how to document it. Conversely happy to just close the PR and someone else can use it to base their own PR on.

Reply:
I looked at the code and the legend indeed knows whether it's an Axes or a Figure legend.

https://github.com/matplotlib/matplotlib/blob/00ec4ec6321d85af8519e268d67807746207f7d2/lib/matplotlib/legend.py#L455-L465

and the `inaxes` parameter is already used for conditional `loc` validation:

https://github.com/matplotlib/matplotlib/blob/00ec4ec6321d85af8519e268d67807746207f7d2/lib/matplotlib/legend.py#L475-L478

I propose it's better to pass `loc` to `Legend` unmodified and have all the validation (and maybe transformation) logic inside.

 I'm sorry I don't have the time to write a PR for that right now. Overall, your PR solves the problem and does not introduce any undesired public API or side-effects. So we may take it as is for now and cleanup the internal logic later. I believe this is needed for better long-term maintainability.

Reply:
OK, I moved it in the newest version.